api/src/main/java/jakarta/json/bind/customization/JsonbCustomization.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package jakarta.json.bind.customization;
+
+import java.text.DateFormat;
+import java.text.NumberFormat;
+import java.util.Optional;
+
+import jakarta.json.bind.adapter.JsonbAdapter;
+import jakarta.json.bind.serializer.JsonbDeserializer;
+
+/**
+ * Common interface for all the customizations.
+ */
+public interface JsonbCustomization {
+
+    /**
+     * Return {@link NumberFormat} specified to the required {@link Scope}.
+     *
+     * @param scope required scope
+     * @return specified {@link NumberFormat} instance, otherwise empty
+     */
+    Optional<NumberFormat> numberFormat(Scope scope);
+
+    /**
+     * Return {@link DateFormat} specified to the required {@link Scope}.
+     *
+     * @param scope required scope
+     * @return specified {@link DateFormat} instance, otherwise empty
+     */
+    Optional<DateFormat> dateFormat(Scope scope);
+
+    /**
+     * Return {@link JsonbDeserializer} of the component.
+     *
+     * @return component deserializer instance, otherwise empty
+     */
+    Optional<JsonbDeserializer<?>> deserializer();
+
+    /**
+     * Return {@link JsonbAdapter} of the component.
+     *
+     * @return component adapter instance, otherwise empty
+     */
+    Optional<JsonbAdapter<?, ?>> adapter();
+
+    /**
+     * Return if the component can be nillable in the given {@link Scope}.
+     *
+     * @param scope required scope
+     * @return property nillable state, otherwise empty
+     */
+    Optional<Boolean> nillable(Scope scope);
```

Comment:
I am not sure the optional makes sense here. If empty, the spec still needs to define whether it is nillable or not when not set.

Reply:
Optional is here because the value didn't have to be set at all. Simply returned boolean is not enough in this case since we need 3 states. True, false and not set

Reply:
Tt is *always* true or false for the runtime.
If you pass empty() you immediately do orElse(true) so at the end it is true or false.
This means you want the user to be able to explicit it or not so means the method is called or not I think, so maybe the API design is reversed compared to the goal and lead to this misleading signature?

Reply:
That is not true. If user didn't call nillable method, it is equivalent to the situation where annotation `@JsonbNillable` has not been applied on the field/method and we have to behave correctly according to that. True/False is just not enough here.

Reply:
@Verdent when @JsonbNillable is not on an attribute the attribute is nillable=true so true/false are strictly sufficient. Now I fully understand the intent of your API which is to let the user say "I don't care, use the (configured or not) default" but then the API is not designed for that case, it is reversed as mentionned.

Reply:
But without the third state you just don't know if this value is intended by user. What if user had set JsonbConfig().withNullValues(true). When checking the customization you wouldn't know if user really wanted to set nillable to false or not. That's the reason for third state. I probably don't understand what you mean that this API is reversed since this covers all the required cases.

Reply:
this is true but if the user did he probably knows too since he has to use the jsonbconfig anyway with that solution so he can wire the default properly without any issue.
It is less misleading to say that it is nillable or not rather than saying it is nillable or not or maybe (still I understand you dont want an enum {TRUE, FALSE, DEFAULT} but optional means something else semantically).

Reply:
What user did? :-) User can set only true or false via builder and not Optional. I don't understand what is misleading here since it just says: User either set true/false by himself or didn't set anything at all since the method has not been called. In that case value from config will be used. There is simply no misleading part here.

Reply:
it is never misleading for the people writing it, this is why reviews are not that bad ;)

javadoc says `Return if the component can be nillable in the given {@link Scope}.`, signature is not compliant to that and in any case you are nillable or not right, whatever you use global config or not so not sure the point to make it indirect for the caller is what I meant.

Reply:
eh... are you trying to tell me... that this whole discussion here was just about incorrect wording? OK, I will change that.

Reply:
hehe, no no, that even the comment is closer to the expectation.
That said I will not fight on these picky details now since I'd like the overall design to be more relevant for end users first so will stick to the main thread.